### PR TITLE
FE: Type Attributes

### PR DESCRIPTION
**Pull request by irenedea**

### CODE REVIEW

**Comment by irenedea** [comment]
Commit 7879193 has caused the following exception when compiling (specifically, I am doing `./gradlew dist`). @demiurg906, do you know what would cause this? It seems to stem from fir calls.
```
Exception in thread "main" java.lang.NoClassDefFoundError: kotlin/reflect/jvm/internal/impl/extensions/ProjectExtensionDescriptor
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
        at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
        at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
        at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at kotlin.reflect.jvm.internal.impl.types.extensions.TypeAttributeTranslators.<clinit>(TypeAttributeTranslators.kt)
        at kotlin.reflect.jvm.internal.impl.serialization.deserialization.DeserializationComponents.<init>(context.kt:58)
        at kotlin.reflect.jvm.internal.impl.load.kotlin.DeserializationComponentsForJava.<init>(DeserializationComponentsForJava.kt:73)
        at kotlin.reflect.jvm.internal.impl.load.kotlin.DeserializationComponentsForJavaKt.makeDeserializationComponentsForJava(DeserializationComponentsForJava.kt:188)
        at kotlin.reflect.jvm.internal.impl.load.kotlin.DeserializationComponentsForJava$Companion.createModuleData(DeserializationComponentsForJava.kt:119)
        at kotlin.reflect.jvm.internal.impl.descriptors.runtime.components.RuntimeModuleData$Companion.create(RuntimeModuleData.kt:32)
        at kotlin.reflect.jvm.internal.ModuleByClassLoaderKt.getOrCreateModule(moduleByClassLoader.kt:58)
        at kotlin.reflect.jvm.internal.KDeclarationContainerImpl$Data$moduleData$2.invoke(KDeclarationContainerImpl.kt:36)
        at kotlin.reflect.jvm.internal.KDeclarationContainerImpl$Data$moduleData$2.invoke(KDeclarationContainerImpl.kt:35)
        at kotlin.reflect.jvm.internal.ReflectProperties$LazySoftVal.invoke(ReflectProperties.java:93)
        at kotlin.reflect.jvm.internal.ReflectProperties$Val.getValue(ReflectProperties.java:32)
        at kotlin.reflect.jvm.internal.KDeclarationContainerImpl$Data.getModuleData(KDeclarationContainerImpl.kt:35)
        at kotlin.reflect.jvm.internal.KClassImpl$Data$descriptor$2.invoke(KClassImpl.kt:50)
        at kotlin.reflect.jvm.internal.KClassImpl$Data$descriptor$2.invoke(KClassImpl.kt:48)
        at kotlin.reflect.jvm.internal.ReflectProperties$LazySoftVal.invoke(ReflectProperties.java:93)
        at kotlin.reflect.jvm.internal.ReflectProperties$Val.getValue(ReflectProperties.java:32)
        at kotlin.reflect.jvm.internal.KClassImpl$Data.getDescriptor(KClassImpl.kt:48)
        at kotlin.reflect.jvm.internal.KClassImpl$Data$supertypes$2.invoke(KClassImpl.kt:125)
        at kotlin.reflect.jvm.internal.KClassImpl$Data$supertypes$2.invoke(KClassImpl.kt:124)
        at kotlin.reflect.jvm.internal.ReflectProperties$LazySoftVal.invoke(ReflectProperties.java:93)
        at kotlin.reflect.jvm.internal.ReflectProperties$Val.getValue(ReflectProperties.java:32)
        at kotlin.reflect.jvm.internal.KClassImpl$Data.getSupertypes(KClassImpl.kt:124)
        at kotlin.reflect.jvm.internal.KClassImpl.getSupertypes(KClassImpl.kt:254)
        at kotlin.reflect.full.KClasses.getAllSupertypes(KClasses.kt:193)
        at kotlin.reflect.full.KClasses.getAllSuperclasses(KClasses.kt:223)
        at org.jetbrains.kotlin.fir.checkers.generator.CheckersConfiguration$1.invoke(CheckersConfiguration.kt:23)
        at org.jetbrains.kotlin.fir.checkers.generator.CheckersConfiguration$1.invoke(CheckersConfiguration.kt:21)
        at org.jetbrains.kotlin.fir.checkers.generator.CheckersConfigurationKt.bfs(CheckersConfiguration.kt:49)
        at org.jetbrains.kotlin.fir.checkers.generator.CheckersConfigurationKt.access$bfs(CheckersConfiguration.kt:1)
        at org.jetbrains.kotlin.fir.checkers.generator.CheckersConfiguration.<init>(CheckersConfiguration.kt:21)
        at org.jetbrains.kotlin.fir.checkers.generator.CheckersConfigurator.build(CheckersConfigurator.kt:30)
        at org.jetbrains.kotlin.fir.checkers.generator.CheckersConfiguratorKt.generateCheckersComponents(CheckersConfigurator.kt:40)
        at org.jetbrains.kotlin.fir.checkers.generator.MainKt.main(Main.kt:23)
Caused by: java.lang.ClassNotFoundException: kotlin.reflect.jvm.internal.impl.extensions.ProjectExtensionDescriptor
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 45 more

```

File: core/descriptors/src/org/jetbrains/kotlin/types/KotlinTypeFactory.kt [hunk 1]
```diff
@@ -157,7 +209,24 @@ object KotlinTypeFactory {
                 if (annotations.isEmpty())
                     it
                 else
-                    AnnotatedSimpleType(it, annotations)
+                    SimpleTypeWithAttributes(it, annotations.toAttributes())
+            }
+
+    @JvmStatic
+    fun simpleTypeWithNonTrivialMemberScope(
```
[the remark this answers is not in the document]
Comment (irenedea) [reply core/descriptors/src/org/jetbrains/kotlin/types/KotlinTypeFactory.kt]: Addressed in 27d9f87

File: compiler/frontend/src/org/jetbrains/kotlin/resolve/calls/inference/BuilderInferenceUtil.kt [hunk 2]
```diff
@@ -56,6 +56,8 @@ class TypeTemplate(
 ) {
     override fun replaceAnnotations(newAnnotations: Annotations) = this
```
[the remark this answers is not in the document]
Comment (irenedea) [reply compiler/frontend/src/org/jetbrains/kotlin/resolve/calls/inference/BuilderInferenceUtil.kt]: Addressed in 
e06bd13

File: compiler/frontend/src/org/jetbrains/kotlin/frontend/di/injection.kt [hunk 3]
```diff
@@ -79,6 +80,8 @@ fun StorageComponentContainer.configureModule(
     analyzerServices.platformConfigurator.configureModuleComponents(this)
     analyzerServices.platformConfigurator.configureModuleDependentCheckers(this)
 
+    useInstance(TypeAttributeTranslators(moduleContext.project))
```
Comment (demiurg906) [inline compiler/frontend/src/org/jetbrains/kotlin/frontend/di/injection.kt]: It's better to use `useImpl<TypeAttributeTranslators>()` here. `project` will be injected automatically

File: core/descriptors/src/org/jetbrains/kotlin/types/CustomAnnotationTypeAttribute.kt [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        if (other == null) return this
+        return CustomAnnotationTypeAttribute(Annotations.create(annotations + other.annotations))
+    }
+
+    override fun isSubtypeOf(other: CustomAnnotationTypeAttribute?): Boolean = true
+
+    override val key: KClass<out CustomAnnotationTypeAttribute>
+        get() = CustomAnnotationTypeAttribute::class
+
+    override fun equals(other: Any?): Boolean {
```
Comment (demiurg906) [inline core/descriptors/src/org/jetbrains/kotlin/types/CustomAnnotationTypeAttribute.kt]: Where is `hashCode`?

File: .idea/artifacts/kotlin_test_wasm_js_1_6_255_SNAPSHOT.xml [hunk 5]
```diff
@@ -1,6 +1,8 @@
 <component name="ArtifactManager">
   <artifact type="jar" name="kotlin-test-wasm-js-1.6.255-SNAPSHOT">
     <output-path>$PROJECT_DIR$/libraries/kotlin.test/wasm/build/libs</output-path>
-    <root id="archive" name="kotlin-test-wasm-js-1.6.255-SNAPSHOT.jar" />
+    <root id="archive" name="kotlin-test-wasm-js-1.6.255-SNAPSHOT.jar">
+      <element id="module-output" name="kotlin.kotlin-test.kotlin-test-wasm.jsMain" />
```
Comment (demiurg906) [inline .idea/artifacts/kotlin_test_wasm_js_1_6_255_SNAPSHOT.xml:5]: Please remove this from commit

File: compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/types/IrTypeSystemContext.kt [hunk 6]
```diff
@@ -386,11 +390,16 @@ interface IrTypeSystemContext : TypeSystemContext, TypeSystemCommonSuperTypesCon
...
 
-    override fun KotlinTypeMarker.getAnnotations(): List<AnnotationMarker> {
+    override fun KotlinTypeMarker.getAttributes(): List<AnnotationMarker> {
         require(this is IrType)
         return this.annotations.map { object : AnnotationMarker, IrElement by it {} }
     }
 
+    override fun KotlinTypeMarker.getCustomAttributes(): List<AnnotationMarker> {
+        require(this is IrType)
+        return emptyList() // TODO: Ir does not currently support custom non-annotation attributes.
```
Comment (demiurg906) [inline compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/types/IrTypeSystemContext.kt:400]: It's not `TODO`. IR backed should not support attributes by design

File: compiler/resolution.common/src/org/jetbrains/kotlin/resolve/calls/inference/model/MutableConstraintStorage.kt [hunk 7]
```diff
@@ -60,7 +60,12 @@ class MutableVariableWithConstraints private constructor(
                 && previousConstraint.isNullabilityConstraint == constraint.isNullabilityConstraint
             ) {
                 if (newConstraintIsUseless(previousConstraint, constraint)) {
-                    return previousConstraint to false
+                    // Preserve constraints with different custom type attributes.
+                    // This allows us to union type attributes in NewCommonSuperTypeCalculator.kt
+                    with(context) {
+                        if (previousConstraint.type.getCustomAttributes() == constraint.type.getCustomAttributes())
```
Comment (demiurg906) [inline compiler/resolution.common/src/org/jetbrains/kotlin/resolve/calls/inference/model/MutableConstraintStorage.kt:66]: Maybe it's better to replace `getCustomAttributes` with `hasCustomAttributes` and only then filter attributes?

**Review by demiurg906 — CHANGES_REQUESTED** [review]

**Comment by demiurg906** [comment]
@irenedea Could you please rebase your branch to the next review iteration and squash most of the commits to cleanup history?

File: core/descriptors/src/org/jetbrains/kotlin/types/TypeSubstitution.kt [hunk 8]
```diff
@@ -144,23 +146,26 @@ fun KotlinType.replace(
 ): KotlinType {
     if ((newArguments.isEmpty() || newArguments === arguments) && newAnnotations === annotations) return this
 
+    val newAttributes = attributes.replaceAnnotations(
+        // Specially handle FilteredAnnotations here due to FilteredAnnotations.isEmpty()
+        if (newAnnotations is FilteredAnnotations && newAnnotations.isEmpty()) Annotations.EMPTY else newAnnotations
```
Comment (irenedea) [inline core/descriptors/src/org/jetbrains/kotlin/types/TypeSubstitution.kt:151]: There is a strange(?) implementation of FilteredAnnotations.isEmpty() that some of the tests rely on: https://github.com/JetBrains/kotlin/blob/master/core/descriptors/src/org/jetbrains/kotlin/descriptors/annotations/Annotations.kt#L71

**Comment by demiurg906** [comment]
Merged manually

**Pull request closed without merging**